contracts/token/LinearVestingWithPenalty.sol

```diff
@@ -457,8 +407,28 @@ contract LinearVestingWithPenalty is Initializable, ERC20Upgradeable, AccessCont
     }
 
     /**
-     * @dev Authorizes upgrades to the contract.
-     * @param newImplementation The address of the new implementation.
-     */
+    * @dev Overrides the internal _update function to transfer vesting schedules
+    * and enforce admin-only access through the onlyAdmin modifier.
+    */
+    function _update(address from, address to, uint256 value) internal virtual override onlyRole(ADMIN_ROLE) {
+        _transferVestingSchedules(from, to); // Streamlined transfer of schedules
+
+        // Call the parent _update function to maintain balance updates and event emissions
+        super._update(from, to, value);
+    }
+
+    /**
+    * @dev Helper function to transfer all vesting schedules from one address to another.
+    * @param from The address to transfer schedules from.
+    * @param to The address to transfer schedules to.
+    */
+    function _transferVestingSchedules(address from, address to) internal {
+        VestingSchedule[] storage schedulesFrom = vestingSchedules[from];
```

Comment:
maybe add a check for the `to` address, like:
```
    if (to == address(0)) revert InvalidRecipientAddress();
```

Reply:
fixed

Reply:
erc 20 will revert  anyway
```
    function _transfer(address from, address to, uint256 value) internal {
        if (from == address(0)) {
            revert ERC20InvalidSender(address(0));
        }
        if (to == address(0)) {
            revert ERC20InvalidReceiver(address(0));
        }
        _update(from, to, value);
    }
    ```

Reply:
ohh, perfect!

Reply:
fixed